explorer/src/db/mod.rs

```diff
@@ -0,0 +1,859 @@
+pub mod error;
+pub mod indexing;
+pub mod multiverse;
+pub mod persistent_sequence;
+mod tally;
+
+use self::error::{BlockNotFound, ExplorerError as Error};
+use self::indexing::{
+    Addresses, Blocks, ChainLengths, EpochData, Epochs, ExplorerAddress, ExplorerBlock,
+    ExplorerVote, ExplorerVotePlan, ExplorerVoteProposal, StakePool, StakePoolBlocks,
+    StakePoolData, Transactions, VotePlans,
+};
+use self::persistent_sequence::PersistentSequence;
+use crate::db::tally::compute_private_tally;
+use crate::db::tally::compute_public_tally;
+use chain_addr::Discrimination;
+use chain_core::property::Block as _;
+use chain_impl_mockchain::{
+    block::HeaderId as HeaderHash,
+    certificate::VotePlanId,
+    stake::{Stake, StakeControl},
+};
+use chain_impl_mockchain::{
+    block::{Block, ChainLength, Epoch},
+    certificate::{Certificate, PoolId},
+    chaintypes::ConsensusVersion,
+    config::ConfigParam,
+    fragment::{ConfigParams, Fragment, FragmentId},
+};
+use chain_impl_mockchain::{fee::LinearFee, vote::PayloadType};
+use futures::prelude::*;
+use multiverse::Multiverse;
+pub use multiverse::Ref;
+use std::convert::Infallible;
+use std::sync::{
+    atomic::{AtomicU32, Ordering},
+    Arc,
+};
+use tokio::sync::{broadcast, RwLock};
+
+#[derive(Clone)]
+pub struct Explorer {
+    pub db: ExplorerDb,
+}
+
+#[derive(Clone)]
+struct Tip(Arc<RwLock<HeaderHash>>);
+
+#[derive(Clone)]
+pub struct ExplorerDb {
+    /// Structure that keeps all the known states to allow easy branch management
+    /// each new block is indexed by getting its previous `State` from the multiverse
+    /// and inserted a new updated one.
+    multiverse: Multiverse,
+    /// This keeps track of the longest chain seen until now. All the queries are
+    /// performed using the state of this branch, the HeaderHash is used as key for the
+    /// multiverse, and the ChainLength is used in the updating process.
+    longest_chain_tip: Tip,
+    pub blockchain_config: BlockchainConfig,
+    stable_store: StableIndex,
+    tip_broadcast: tokio::sync::broadcast::Sender<(HeaderHash, multiverse::Ref)>,
+}
+
+#[derive(Clone)]
+pub struct StableIndex {
+    confirmed_block_chain_length: Arc<AtomicU32>,
+}
+
+#[derive(Clone)]
+pub struct BlockchainConfig {
+    /// Used to construct `Address` from `AccountIndentifier` when processing transaction
+    /// inputs
+    pub discrimination: Discrimination,
+    pub consensus_version: ConsensusVersion,
+    pub fees: LinearFee,
+    pub epoch_stability_depth: u32,
+}
+
+/// Inmutable data structure used to represent the explorer's state at a given Block
+/// A new state can be obtained to from a Block and it's previous state, getting two
+/// independent states but with memory sharing to minimize resource utilization
+#[derive(Clone)]
+pub struct State {
+    pub transactions: Transactions,
+    pub blocks: Blocks,
+    addresses: Addresses,
+    epochs: Epochs,
+    chain_lengths: ChainLengths,
```

Comment:
are this only the chain lengths of the ancestors of the block corresponding to this state?

Reply:
it's a reverse lookup for `ChainLength` -> `BlockHash`. And yes, although it should also include the hash of the block that produced this `State`.